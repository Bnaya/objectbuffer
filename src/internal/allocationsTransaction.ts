import { IMemPool } from "@bnaya/malloc-temporary-fork";
import { OutOfMemoryError } from "./exceptions";

// extend pool and not monkey patch? need to think about it
export function allocationsTransaction(operation: () => void, pool: IMemPool) {
  const allocations: number[] = [];
  const originalMalloc = pool.malloc;
  const originalCalloc = pool.calloc;
  const originalRealloc = pool.realloc;
  let isInRealloc = false;

  function rollback() {
    for (const block of allocations) {
      pool.free(block);
    }
  }

  pool.realloc = function realloc(ptr: number, size: number) {
    isInRealloc = true;
    const allocation = originalRealloc.call(pool, ptr, size);

    if (allocation === 0) {
      throw new OutOfMemoryError();
    }

    isInRealloc = false;

    return allocation;
  };

  pool.malloc = function malloc(size: number) {
    const allocation = originalMalloc.call(pool, size);

    if (allocation === 0) {
      rollback();
      throw new OutOfMemoryError();
    }

    if (!isInRealloc) {
      allocations.push(allocation);
    }

    return allocation;
  };

  pool.calloc = function calloc(size: number) {
    const allocation = originalCalloc.call(pool, size);

    if (allocation === 0) {
      rollback();
      throw new OutOfMemoryError();
    }

    if (!isInRealloc) {
      allocations.push(allocation);
    }

    return allocation;
  };

  operation();

  pool.malloc = originalMalloc;
  pool.calloc = originalCalloc;
  pool.realloc = originalRealloc;
}
