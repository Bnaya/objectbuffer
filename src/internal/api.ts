import { initializeArrayBuffer, incrementRefCount } from "./store";
import { createObjectWrapper } from "./objectWrapper";
import type {
  ObjectBufferSettings,
  GlobalCarrier,
  MemoryStats,
} from "./interfaces";
import {
  arrayBufferCopyTo,
  externalArgsApiToExternalArgsApi,
  getInternalAPI,
  isSupportedTopLevelValue,
  getEndiannessOfSystem,
} from "./utils";
import {
  INITIAL_ENTRY_POINTER_TO_POINTER,
  MEM_POOL_START,
  ENDIANNESS_FLAG_POINTER,
} from "./consts";
import { OutOfMemoryError, UnsupportedOperationError } from "./exceptions";
import { saveValueIterative } from "./saveValue";
import { TransactionalAllocator } from "./TransactionalAllocator";
import { freeNoLongerUsedAddresses } from "./freeNoLongerUsedAddresses";

/**
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#Browser_compatibility
 *
 * Future holds additional types https://github.com/tc39/proposal-resizablearraybuffer
 */
export type ArrayBufferKind = "vanilla" | "shared";

/**
 * Create a new objectBuffer, with the given initial value
 * @param size The size of the ArrayBuffer to create (heap size)
 * @param initialValue
 * @param settings
 * @param arrayBufferConstructor
 */
export function createObjectBuffer<T = any>(
  size: number,
  initialValue: T,
  settings: ObjectBufferSettings = {},
  arrayBufferKind: ArrayBufferKind = "vanilla"
): T {
  if (!isSupportedTopLevelValue(initialValue)) {
    throw new UnsupportedOperationError();
  }

  const allocator = new TransactionalAllocator(
    {
      start: MEM_POOL_START,
      size,
    },
    arrayBufferKind === "shared"
  );

  const arrayBuffer = allocator.getArrayBuffer();

  initializeArrayBuffer(arrayBuffer);

  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const referencedPointers: number[] = [];

  allocator.transaction(() => {
    saveValueIterative(
      externalArgsApiToExternalArgsApi(settings),
      carrier,
      referencedPointers,
      INITIAL_ENTRY_POINTER_TO_POINTER,
      initialValue
    );
  });

  for (const pointer of referencedPointers) {
    incrementRefCount(carrier.heap, pointer);
  }

  const dv = new DataView(arrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  dv.setUint32(ENDIANNESS_FLAG_POINTER, getEndiannessOfSystem(), true);

  return createObjectWrapper(
    externalArgsApiToExternalArgsApi(settings),
    carrier,
    carrier.heap.u32[
      INITIAL_ENTRY_POINTER_TO_POINTER / Uint32Array.BYTES_PER_ELEMENT
    ]
  );
}

/**
 * Grow or shrink the underlying ArrayBuffer
 *
 * @unstable
 * Due to possible issues with future support of typed arrays,
 * and the upcoming proposal, this api function may be removed
 * https://github.com/tc39/proposal-resizablearraybuffer
 *
 * @param objectBuffer
 * @param newSize
 *
 */
export function unstable_resizeObjectBuffer(
  objectBuffer: unknown,
  newSize: number
) {
  if (newSize < memoryStats(objectBuffer).top) {
    throw new OutOfMemoryError();
  }

  const oldArrayBuffer = getUnderlyingArrayBuffer(objectBuffer);
  const newArrayBuffer = new ArrayBuffer(newSize);

  arrayBufferCopyTo(
    oldArrayBuffer,
    0,
    Math.min(newSize, oldArrayBuffer.byteLength),
    newArrayBuffer,
    0
  );

  unstable_replaceUnderlyingArrayBuffer(objectBuffer, newArrayBuffer);

  return newArrayBuffer;
}

export function getUnderlyingArrayBuffer(
  objectBuffer: unknown
): ArrayBuffer | SharedArrayBuffer {
  return getInternalAPI(objectBuffer).getCarrier().heap.u8.buffer;
}

/**
 * Create objectBuffer object from the given ArrayBuffer
 *
 * The given ArrayBuffer is expected to be one obtained via getUnderlyingArrayBuffer
 * This operation doesn't change any value in the ArrayBuffer
 *
 * @param settings
 * @param arrayBuffer
 */
export function loadObjectBuffer<T = any>(
  arrayBuffer: ArrayBuffer | SharedArrayBuffer,
  settings: ObjectBufferSettings = {}
): T {
  const allocator = TransactionalAllocator.load(arrayBuffer);

  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const dv = new DataView(arrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  const endiannessOfGivenAb = dv.getUint32(ENDIANNESS_FLAG_POINTER, true);

  if (endiannessOfGivenAb !== getEndiannessOfSystem()) {
    throw new Error("Endianness mismatch");
  }

  return createObjectWrapper(
    externalArgsApiToExternalArgsApi(settings),
    carrier,
    carrier.heap.u32[
      INITIAL_ENTRY_POINTER_TO_POINTER / Uint32Array.BYTES_PER_ELEMENT
    ]
  );
}

/**
 * Replace the Underlying array buffer with the given one.
 * The given ArrayBuffer is expected to be a copy of the prev ArrayBuffer,
 * just bigger or smaller (less free space)
 *
 * Consider using `resizeObjectBuffer`
 *
 * @unstable
 * Due to possible issues with future support of typed arrays,
 * and the upcoming proposal, this api function may be removed
 * https://github.com/tc39/proposal-resizablearraybuffer
 *
 * @param objectBuffer
 * @param newArrayBuffer
 */
export function unstable_replaceUnderlyingArrayBuffer(
  objectBuffer: unknown,
  newArrayBuffer: ArrayBuffer | SharedArrayBuffer
) {
  const allocator = TransactionalAllocator.load(newArrayBuffer);
  const carrier: GlobalCarrier = {
    allocator,
    heap: allocator.getHeap(),
  };

  const dv = new DataView(newArrayBuffer);
  // endianness flag is always saved in little endian so we can read in every system endianness
  const endiannessOfGivenAb = dv.getUint32(ENDIANNESS_FLAG_POINTER, true);

  if (endiannessOfGivenAb !== getEndiannessOfSystem()) {
    throw new Error("Endianness mismatch");
  }

  allocator.setNewEnd(newArrayBuffer.byteLength);

  getInternalAPI(objectBuffer).replaceCarrierContent(carrier);
}

/**
 * Return the number of free & used bytes left in the given objectBuffer
 */
export function memoryStats(objectBuffer: unknown): MemoryStats {
  const { available, total, top } = getInternalAPI(objectBuffer)
    .getCarrier()
    .allocator.stats();

  return { available, used: total - available, total, top };
}

export { reclaim, queueReclaim } from "./reclaim";

/**
 *  Update the settings of the given ObjectBuffer
 */
export function updateObjectBufferSettings(
  objectBuffer: unknown,
  options: ObjectBufferSettings
) {
  Object.assign(getInternalAPI(objectBuffer).getExternalArgs(), options);
}

export function readObjectBufferSettings(objectBuffer: unknown) {
  return getInternalAPI(objectBuffer).getExternalArgs();
}

/**
 * Free all the addresses collected using FinalizationRegistry
 * When no FinalizationRegistry/WeakRef available, use disposeWrapperObject
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry
 *
 * This is not called automatic by FinalizationRegistry,
 * because It's only safe to call it when you have a lock/similar (As any other operation)
 * And FinalizationRegistry might run when ever
 */
export function processQueuedReclaims(objectBuffer: unknown) {
  freeNoLongerUsedAddresses(getInternalAPI(objectBuffer).getCarrier());
}
