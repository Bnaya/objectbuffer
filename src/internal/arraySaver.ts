import { appendEntry } from "./store";
import { ENTRY_TYPE } from "./entry-types";
import { ArrayEntry, ExternalArgs, GlobalCarrier } from "./interfaces";
import { saveValue } from "./saveValue";

export function arraySaver(
  externalArgs: ExternalArgs,
  carrier: GlobalCarrier,
  referencedPointers: number[],
  arrayToSave: Array<any>
) {
  let memoryForPointersCursor = carrier.allocator.calloc(
    (arrayToSave.length + externalArgs.arrayAdditionalAllocation) *
      Uint32Array.BYTES_PER_ELEMENT
  );

  const arrayStartEntry: ArrayEntry = {
    type: ENTRY_TYPE.ARRAY,
    refsCount: 1,
    value: memoryForPointersCursor,
    allocatedLength:
      arrayToSave.length + externalArgs.arrayAdditionalAllocation,
    length: arrayToSave.length
  };

  for (const item of arrayToSave) {
    const rOfValue = saveValue(externalArgs, carrier, referencedPointers, item);

    carrier.uint32[
      memoryForPointersCursor / Uint32Array.BYTES_PER_ELEMENT
    ] = rOfValue;

    memoryForPointersCursor += Uint32Array.BYTES_PER_ELEMENT;
  }

  const arrayEntryAppendResult = appendEntry(
    externalArgs,
    carrier,
    arrayStartEntry
  );

  return arrayEntryAppendResult;
}
