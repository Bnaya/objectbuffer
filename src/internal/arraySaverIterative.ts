import { ENTRY_TYPE } from "./entry-types";
import { GlobalCarrier } from "./interfaces";
import { array_set_all, array_size } from "./generatedStructs";

export function arraySaverIterative(
  arrayAdditionalAllocation: number,
  carrier: GlobalCarrier,
  valuesToSave: unknown[],
  pointersToSaveTo: number[],
  arrayToSave: unknown[]
) {
  const arrayLength = arrayToSave.length;

  const arrayStructPointer = carrier.allocator.calloc(array_size);
  const arrayPointersSpaceStart = carrier.allocator.calloc(
    (arrayLength + arrayAdditionalAllocation) * Uint32Array.BYTES_PER_ELEMENT
  );

  array_set_all(
    carrier.heap,
    arrayStructPointer,
    ENTRY_TYPE.ARRAY,
    // ARC starts with 1
    1,
    arrayPointersSpaceStart,
    arrayLength,
    arrayLength + arrayAdditionalAllocation
  );

  for (let i = 0; i < arrayLength; i += 1) {
    valuesToSave.push(arrayToSave[i]);
    pointersToSaveTo.push(
      arrayPointersSpaceStart + i * Uint32Array.BYTES_PER_ELEMENT
    );
  }

  return arrayStructPointer;
}
