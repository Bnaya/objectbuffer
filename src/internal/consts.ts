export const GLOBAL_LOCK_OFFSET = 0;
export const INITIAL_ENTRY_POINTER_TO_POINTER =
  GLOBAL_LOCK_OFFSET + Uint32Array.BYTES_PER_ELEMENT;

export const ENDIANNESS_FLAG_POINTER =
  INITIAL_ENTRY_POINTER_TO_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export enum ENDIANNESS {
  BIG,
  LITTLE,
}

export const TEMP_SAVE_POINTER =
  ENDIANNESS_FLAG_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export const INITIAL_ENTRY_POINTER_VALUE =
  TEMP_SAVE_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export const MEM_POOL_START =
  INITIAL_ENTRY_POINTER_VALUE + Uint32Array.BYTES_PER_ELEMENT;

export const UNDEFINED_KNOWN_ADDRESS = 0;
export const NULL_KNOWN_ADDRESS = 1;
export const TRUE_KNOWN_ADDRESS = 2;
export const FALSE_KNOWN_ADDRESS = 3;
export const MAX_64_BIG_INT = BigInt("0xFFFFFFFFFFFFFFFF");
