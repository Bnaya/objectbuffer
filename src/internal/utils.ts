import type {
  ExternalArgs,
  InternalAPI,
  ObjectBufferSettings,
} from "./interfaces";
import { ENTRY_TYPE } from "./entry-types";
import { INTERNAL_API_SYMBOL } from "./symbols";
import {
  UNDEFINED_KNOWN_ADDRESS,
  NULL_KNOWN_ADDRESS,
  TRUE_KNOWN_ADDRESS,
  FALSE_KNOWN_ADDRESS,
  ENDIANNESS,
} from "./consts";
import { FunctionalAllocatorWrapper } from "./TransactionalAllocator";

export function getEndiannessOfSystem() {
  const F64 = new Float64Array(1);
  const U32 = new Uint32Array(F64.buffer);
  F64[0] = 2;
  if (U32[1] === 0x40000000) {
    return ENDIANNESS.LITTLE;
  } else {
    return ENDIANNESS.BIG;
  }
}

export function createKnownTypeGuard<T>(arr: ReadonlyArray<T>) {
  return function knownTypeGuard(v: unknown): v is T {
    return arr.includes(v as any);
  };
}

export function invariant(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(message);
  }
}

export function arrayBufferCopyTo(
  origin: ArrayBuffer,
  startByte: number,
  length: number,
  target: ArrayBuffer,
  toTargetByte: number
) {
  const copyFrom = new Uint8Array(origin);
  const copyTo = new Uint8Array(target);

  copyTo.set(copyFrom.subarray(startByte, startByte + length), toTargetByte);
}

export function getOurPointerIfApplicable(
  value: any,
  ourAllocator: FunctionalAllocatorWrapper
) {
  if (INTERNAL_API_SYMBOL in value) {
    const api = getInternalAPI(value);
    if (api.getCarrier().allocator === ourAllocator) {
      return api.getEntryPointer();
    }
  }
}

export function externalArgsApiToExternalArgsApi(
  p: ObjectBufferSettings
): ExternalArgs {
  return {
    ...p,
    hashMapMinInitialCapacity: p.hashMapMinInitialCapacity
      ? p.hashMapMinInitialCapacity
      : 8,
    hashMapLoadFactor: p.hashMapLoadFactor ? p.hashMapLoadFactor : 0.75,
    arrayAdditionalAllocation: p.arrayAdditionalAllocation
      ? p.arrayAdditionalAllocation
      : 0,
  };
}

export function getInternalAPI(value: any): InternalAPI {
  if (!value[INTERNAL_API_SYMBOL]) {
    throw new RangeError("getInternalAPI not applicable");
  }

  return value[INTERNAL_API_SYMBOL];
}

export function align(value: number, alignTo = 8) {
  return Math.ceil(value / alignTo) * alignTo;
}

export function isKnownAddressValuePointer(entryPointer: number) {
  return (
    entryPointer === UNDEFINED_KNOWN_ADDRESS ||
    entryPointer === NULL_KNOWN_ADDRESS ||
    entryPointer === TRUE_KNOWN_ADDRESS ||
    entryPointer === FALSE_KNOWN_ADDRESS
  );
}

export function isTypeWithRC(type: ENTRY_TYPE) {
  return (
    type === ENTRY_TYPE.OBJECT ||
    type === ENTRY_TYPE.ARRAY ||
    type === ENTRY_TYPE.DATE ||
    type === ENTRY_TYPE.MAP ||
    type === ENTRY_TYPE.SET ||
    type === ENTRY_TYPE.STRING
  );
}

/**
 *
 * I hope It's reliable
 */
export function isSupportedTopLevelValue(value: unknown) {
  return !(
    Array.isArray(value) ||
    value instanceof Date ||
    value instanceof Map ||
    value instanceof Set ||
    typeof value !== "object" ||
    typeof value === null ||
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    value!.constructor.name !== "Object"
  );
}

export function fromEntries(iterable: Iterable<unknown>): unknown {
  // @ts-expect-error we don't care
  return [...iterable].reduce(function (obj: any, [key, val]) {
    obj[key] = val;
    return obj;
  }, {});
}
