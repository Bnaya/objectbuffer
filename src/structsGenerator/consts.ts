import { fromEntries } from "../internal/utils";

export const typedArraysPropNameToCtorMap = {
  u8: Uint8Array,
  u8c: Uint8ClampedArray,
  i8: Int8Array,
  u16: Uint16Array,
  i16: Int16Array,
  u32: Uint32Array,
  i32: Int32Array,
  f32: Float32Array,
  f64: Float64Array,
  b64: BigInt64Array,
  u64: BigUint64Array,
} as const;

export type TypedArrayPropNameToCtorType = typeof typedArraysPropNameToCtorMap;

const intermediate1 = Object.entries(typedArraysPropNameToCtorMap).map(
  ([key, value]) => {
    return [value.name, key];
  }
);

export const typedArrayNameToHeapProp: any = fromEntries(intermediate1);

export type Heap = {
  [x in keyof TypedArrayPropNameToCtorType]: InstanceType<
    TypedArrayPropNameToCtorType[x]
  >;
};

export type PossibleTypedArrays =
  TypedArrayPropNameToCtorType[keyof TypedArrayPropNameToCtorType];
export type StructManifest = { [x: string]: PossibleTypedArrays };

export function createHeap(sb: ArrayBuffer | SharedArrayBuffer): Heap {
  const r = Object.fromEntries(
    Object.entries(typedArraysPropNameToCtorMap).map(([name, Ctor]) => [
      name,
      new Ctor(sb),
    ])
  );

  return r as any;
}
