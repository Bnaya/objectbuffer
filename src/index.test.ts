/* eslint-env jest */

import {
  createObjectBuffer,
  getUnderlyingArrayBuffer,
  loadObjectBuffer,
} from ".";
import {
  arrayBuffer2HexArray,
  getArrayBufferOnTopSize,
} from "./internal/testUtils";
import { externalArgsApiToExternalArgsApi } from "./internal/utils";
import { ENDIANNESS_FLAG_POINTER, ENDIANNESS } from "./internal/consts";

describe("createObjectBuffer", () => {
  test("createObjectBuffer simple", () => {
    const o = createObjectBuffer(1024, {
      a: "b",
      b: null,
      c: { t: 5 },
    });

    expect(o).toMatchInlineSnapshot(`
      Object {
        "a": "b",
        "b": null,
        "c": Object {
          "t": 5,
        },
      }
    `);
  });
});

describe("getUnderlyingArrayBuffer", () => {
  const externalArgs = externalArgsApiToExternalArgsApi({
    arrayAdditionalAllocation: 0,
  });
  test("getUnderlyingArrayBuffer simple", () => {
    const o = createObjectBuffer(
      1024,
      {
        b: null,
        c: { t: 5 },
      },
      externalArgs
    );

    const arrayBuffer = getUnderlyingArrayBuffer(o);

    expect(arrayBuffer).toBeInstanceOf(ArrayBuffer);

    // that one tests implementation details, but...
    expect(
      arrayBuffer2HexArray(getArrayBufferOnTopSize(o), true)
    ).toMatchSnapshot();
  });
});

describe("loadObjectBuffer", () => {
  test("loadObjectBuffer simple", () => {
    const o = createObjectBuffer(1024, {
      a: "b",
      b: null,
      c: { t: 5 },
    });

    const arrayBuffer = getUnderlyingArrayBuffer(o);

    const dv = new DataView(arrayBuffer);

    const realEndianness = dv.getUint32(ENDIANNESS_FLAG_POINTER, true);
    // flip Endianness
    dv.setUint32(
      ENDIANNESS_FLAG_POINTER,
      realEndianness === ENDIANNESS.BIG ? ENDIANNESS.LITTLE : ENDIANNESS.BIG,
      true
    );

    expect(() => {
      return loadObjectBuffer(arrayBuffer);
    }).toThrowErrorMatchingInlineSnapshot(`"Endianness mismatch"`);
  });

  test("Endianness miss match", () => {
    const o = createObjectBuffer(1024, {
      a: "b",
      b: null,
      c: { t: 5 },
    });

    const arrayBuffer = getUnderlyingArrayBuffer(o);

    const newOne = loadObjectBuffer(arrayBuffer);

    expect(o).toMatchInlineSnapshot(`
      Object {
        "a": "b",
        "b": null,
        "c": Object {
          "t": 5,
        },
      }
    `);

    expect(newOne).toMatchInlineSnapshot(`
      Object {
        "a": "b",
        "b": null,
        "c": Object {
          "t": 5,
        },
      }
    `);
  });
});
