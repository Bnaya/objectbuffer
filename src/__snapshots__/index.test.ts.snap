// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUnderlyingArrayBuffer getUnderlyingArrayBuffer simple 1`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x38",
  "5:0x00",
  "6:0x00",
  "7:0x00",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0x00",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0x00",
  "17:0x00",
  "18:0x00",
  "19:0x00",
  "20:0x28",
  "21:0x02",
  "22:0x00",
  "23:0x00",
  "24:0x40",
  "25:0x02",
  "26:0x00",
  "27:0x00",
  "28:0x00",
  "29:0x04",
  "30:0x00",
  "31:0x00",
  "32:0x08",
  "33:0x00",
  "34:0x00",
  "35:0x00",
  "36:0x03",
  "37:0x00",
  "38:0x00",
  "39:0x00",
  "40:0x10",
  "41:0x00",
  "42:0x00",
  "43:0x00",
  "44:0x00",
  "45:0x00",
  "46:0x00",
  "47:0x00",
  "48:0x18",
  "49:0x00",
  "50:0x00",
  "51:0x00",
  "52:0x00",
  "53:0x00",
  "54:0x00",
  "55:0x00",
  "56:0x00",
  "57:0x00",
  "58:0x00",
  "59:0x00",
  "60:0x00",
  "61:0x00",
  "62:0x1c",
  "63:0x40",
  "64:0x02",
  "65:0x00",
  "66:0x00",
  "67:0x00",
  "68:0x50",
  "69:0x00",
  "70:0x00",
  "71:0x00",
  "72:0x18",
  "73:0x00",
  "74:0x00",
  "75:0x00",
  "76:0x30",
  "77:0x00",
  "78:0x00",
  "79:0x00",
  "80:0x68",
  "81:0x00",
  "82:0x00",
  "83:0x00",
  "84:0x90",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x02",
  "89:0x00",
  "90:0x00",
  "91:0x00",
  "92:0x08",
  "93:0x02",
  "94:0x00",
  "95:0x00",
  "96:0x28",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x48",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x00",
  "105:0x00",
  "106:0x00",
  "107:0x00",
  "108:0x00",
  "109:0x00",
  "110:0x00",
  "111:0x00",
  "112:0xe0",
  "113:0x00",
  "114:0x00",
  "115:0x00",
  "116:0x38",
  "117:0x01",
  "118:0x00",
  "119:0x00",
  "120:0x00",
  "121:0x00",
  "122:0x00",
  "123:0x00",
  "124:0x00",
  "125:0x00",
  "126:0x00",
  "127:0x00",
  "128:0x00",
  "129:0x00",
  "130:0x00",
  "131:0x00",
  "132:0x00",
  "133:0x00",
  "134:0x00",
  "135:0x00",
  "136:0x10",
  "137:0x00",
  "138:0x00",
  "139:0x00",
  "140:0x60",
  "141:0x00",
  "142:0x00",
  "143:0x00",
  "144:0x50",
  "145:0x01",
  "146:0x00",
  "147:0x00",
  "148:0xa0",
  "149:0x00",
  "150:0x00",
  "151:0x00",
  "152:0x10",
  "153:0x00",
  "154:0x00",
  "155:0x00",
  "156:0x88",
  "157:0x00",
  "158:0x00",
  "159:0x00",
  "160:0xf8",
  "161:0x00",
  "162:0x00",
  "163:0x00",
  "164:0xe0",
  "165:0x00",
  "166:0x00",
  "167:0x00",
  "168:0x10",
  "169:0x00",
  "170:0x00",
  "171:0x00",
  "172:0x98",
  "173:0x00",
  "174:0x00",
  "175:0x00",
  "176:0x62",
  "177:0x00",
  "178:0x00",
  "179:0x00",
  "180:0x00",
  "181:0x00",
  "182:0x00",
  "183:0x00",
  "184:0x20",
  "185:0x00",
  "186:0x00",
  "187:0x00",
  "188:0xa8",
  "189:0x00",
  "190:0x00",
  "191:0x00",
  "192:0x00",
  "193:0x00",
  "194:0x00",
  "195:0x00",
  "196:0x00",
  "197:0x00",
  "198:0x14",
  "199:0x40",
  "200:0x01",
  "201:0x00",
  "202:0x00",
  "203:0x00",
  "204:0x01",
  "205:0x00",
  "206:0x00",
  "207:0x00",
  "208:0xb0",
  "209:0x00",
  "210:0x00",
  "211:0x00",
  "212:0x00",
  "213:0x00",
  "214:0x00",
  "215:0x00",
  "216:0x18",
  "217:0x00",
  "218:0x00",
  "219:0x00",
  "220:0xb8",
  "221:0x00",
  "222:0x00",
  "223:0x00",
  "224:0x01",
  "225:0x00",
  "226:0x00",
  "227:0x00",
  "228:0x00",
  "229:0x00",
  "230:0x00",
  "231:0x00",
  "232:0xc0",
  "233:0x00",
  "234:0x00",
  "235:0x00",
  "236:0xa0",
  "237:0x00",
  "238:0x00",
  "239:0x00",
  "240:0x10",
  "241:0x00",
  "242:0x00",
  "243:0x00",
  "244:0xd8",
  "245:0x00",
  "246:0x00",
  "247:0x00",
  "248:0x50",
  "249:0x01",
  "250:0x00",
  "251:0x00",
  "252:0x38",
  "253:0x01",
  "254:0x00",
  "255:0x00",
  "256:0x10",
  "257:0x00",
  "258:0x00",
  "259:0x00",
  "260:0xf0",
  "261:0x00",
  "262:0x00",
  "263:0x00",
  "264:0x63",
  "265:0x00",
  "266:0x00",
  "267:0x00",
  "268:0x00",
  "269:0x00",
  "270:0x00",
  "271:0x00",
  "272:0x20",
  "273:0x00",
  "274:0x00",
  "275:0x00",
  "276:0x00",
  "277:0x01",
  "278:0x00",
  "279:0x00",
  "280:0x00",
  "281:0x00",
  "282:0x00",
  "283:0x00",
  "284:0x00",
  "285:0x00",
  "286:0x14",
  "287:0x40",
  "288:0x01",
  "289:0x00",
  "290:0x00",
  "291:0x00",
  "292:0x01",
  "293:0x00",
  "294:0x00",
  "295:0x00",
  "296:0x08",
  "297:0x01",
  "298:0x00",
  "299:0x00",
  "300:0x00",
  "301:0x00",
  "302:0x00",
  "303:0x00",
  "304:0x18",
  "305:0x00",
  "306:0x00",
  "307:0x00",
  "308:0x10",
  "309:0x01",
  "310:0x00",
  "311:0x00",
  "312:0x60",
  "313:0x01",
  "314:0x00",
  "315:0x00",
  "316:0x00",
  "317:0x00",
  "318:0x00",
  "319:0x00",
  "320:0x18",
  "321:0x01",
  "322:0x00",
  "323:0x00",
  "324:0xf8",
  "325:0x00",
  "326:0x00",
  "327:0x00",
  "328:0x10",
  "329:0x00",
  "330:0x00",
  "331:0x00",
  "332:0x30",
  "333:0x01",
  "334:0x00",
  "335:0x00",
  "336:0x00",
  "337:0x00",
  "338:0x00",
  "339:0x00",
  "340:0x00",
  "341:0x00",
  "342:0x00",
  "343:0x00",
  "344:0x18",
  "345:0x00",
  "346:0x00",
  "347:0x00",
  "348:0x48",
  "349:0x01",
  "350:0x00",
  "351:0x00",
  "352:0x00",
  "353:0x00",
  "354:0x00",
  "355:0x00",
  "356:0x00",
  "357:0x00",
  "358:0x1c",
  "359:0x40",
  "360:0x01",
  "361:0x00",
  "362:0x00",
  "363:0x00",
  "364:0x78",
  "365:0x01",
  "366:0x00",
  "367:0x00",
  "368:0x18",
  "369:0x00",
  "370:0x00",
  "371:0x00",
  "372:0x58",
  "373:0x01",
  "374:0x00",
  "375:0x00",
  "376:0x90",
  "377:0x01",
  "378:0x00",
  "379:0x00",
  "380:0xb8",
  "381:0x01",
  "382:0x00",
  "383:0x00",
  "384:0x01",
  "385:0x00",
  "386:0x00",
  "387:0x00",
  "388:0x08",
  "389:0x01",
  "390:0x00",
  "391:0x00",
  "392:0x28",
  "393:0x00",
  "394:0x00",
  "395:0x00",
  "396:0x70",
  "397:0x01",
  "398:0x00",
  "399:0x00",
  "400:0x00",
  "401:0x00",
  "402:0x00",
  "403:0x00",
  "404:0x00",
  "405:0x00",
  "406:0x00",
  "407:0x00",
  "408:0x00",
  "409:0x00",
  "410:0x00",
  "411:0x00",
  "412:0x00",
  "413:0x00",
  "414:0x00",
  "415:0x00",
  "416:0x08",
  "417:0x02",
  "418:0x00",
  "419:0x00",
  "420:0x00",
  "421:0x00",
  "422:0x00",
  "423:0x00",
  "424:0x00",
  "425:0x00",
  "426:0x00",
  "427:0x00",
  "428:0x00",
  "429:0x00",
  "430:0x00",
  "431:0x00",
  "432:0x10",
  "433:0x00",
  "434:0x00",
  "435:0x00",
  "436:0x88",
  "437:0x01",
  "438:0x00",
  "439:0x00",
  "440:0x20",
  "441:0x02",
  "442:0x00",
  "443:0x00",
  "444:0xc8",
  "445:0x01",
  "446:0x00",
  "447:0x00",
  "448:0x10",
  "449:0x00",
  "450:0x00",
  "451:0x00",
  "452:0xb0",
  "453:0x01",
  "454:0x00",
  "455:0x00",
  "456:0x20",
  "457:0x02",
  "458:0x00",
  "459:0x00",
  "460:0x08",
  "461:0x02",
  "462:0x00",
  "463:0x00",
  "464:0x10",
  "465:0x00",
  "466:0x00",
  "467:0x00",
  "468:0xc0",
  "469:0x01",
  "470:0x00",
  "471:0x00",
  "472:0x74",
  "473:0x00",
  "474:0x00",
  "475:0x00",
  "476:0x00",
  "477:0x00",
  "478:0x00",
  "479:0x00",
  "480:0x20",
  "481:0x00",
  "482:0x00",
  "483:0x00",
  "484:0xd0",
  "485:0x01",
  "486:0x00",
  "487:0x00",
  "488:0x00",
  "489:0x00",
  "490:0x00",
  "491:0x00",
  "492:0x00",
  "493:0x00",
  "494:0x14",
  "495:0x40",
  "496:0x01",
  "497:0x00",
  "498:0x00",
  "499:0x00",
  "500:0x01",
  "501:0x00",
  "502:0x00",
  "503:0x00",
  "504:0xd8",
  "505:0x01",
  "506:0x00",
  "507:0x00",
  "508:0x00",
  "509:0x00",
  "510:0x00",
  "511:0x00",
  "512:0x18",
  "513:0x00",
  "514:0x00",
  "515:0x00",
  "516:0xe0",
  "517:0x01",
  "518:0x00",
  "519:0x00",
  "520:0x30",
  "521:0x02",
  "522:0x00",
  "523:0x00",
  "524:0x00",
  "525:0x00",
  "526:0x00",
  "527:0x00",
  "528:0xe8",
  "529:0x01",
  "530:0x00",
  "531:0x00",
  "532:0xc8",
  "533:0x01",
  "534:0x00",
  "535:0x00",
  "536:0x10",
  "537:0x00",
  "538:0x00",
  "539:0x00",
  "540:0x00",
  "541:0x02",
  "542:0x00",
  "543:0x00",
  "544:0x00",
  "545:0x00",
  "546:0x00",
  "547:0x00",
  "548:0x00",
  "549:0x00",
  "550:0x00",
  "551:0x00",
  "552:0x18",
  "553:0x00",
  "554:0x00",
  "555:0x00",
  "556:0x18",
  "557:0x02",
  "558:0x00",
  "559:0x00",
  "560:0x00",
  "561:0x00",
  "562:0x00",
  "563:0x00",
  "564:0x00",
  "565:0x00",
  "566:0x00",
  "567:0x40",
  "568:0x00",
  "569:0x00",
  "570:0x00",
  "571:0x00",
  "572:0x00",
  "573:0x00",
  "574:0x14",
  "575:0x40",
]
`;
